src/common/virtual_file_system.cpp

```diff
@@ -216,16 +218,67 @@ bool VirtualFileSystem::SubSystemIsDisabled(const string &name) {
 	return disabled_file_systems.find(name) != disabled_file_systems.end();
 }
 
+FileSystem &VirtualFileSystem::FindFileSystem(const string &path, optional_ptr<FileOpener> opener) {
+	return FindFileSystem(path, FileOpener::TryGetClientContext(opener));
+}
+
+FileSystem &VirtualFileSystem::FindFileSystem(const string &path, optional_ptr<ClientContext> client_context) {
+	bool should_autoload = true;
+	auto &fs = FindFileSystemInternal(path, should_autoload);
+
+	if (should_autoload && client_context) {
+		string required_extension;
+
+		for (const auto &entry : EXTENSION_FILE_PREFIXES) {
+			if (StringUtil::StartsWith(path, entry.name)) {
+				required_extension = entry.extension;
+			}
+		}
+		if (!required_extension.empty() && client_context &&
+		    !client_context->db->ExtensionIsLoaded(required_extension)) {
+			auto &dbconfig = DBConfig::GetConfig(*client_context);
+			if (!ExtensionHelper::CanAutoloadExtension(required_extension) ||
+			    !dbconfig.options.autoload_known_extensions) {
+				auto error_message = "File " + path + " requires the extension " + required_extension + " to be loaded";
+				error_message = ExtensionHelper::AddExtensionInstallHintToErrorMsg(*client_context, error_message,
+				                                                                   required_extension);
+				throw MissingExtensionException(error_message);
+			}
+			// an extension is required to read this file, but it is not loaded - try to load it
+			ExtensionHelper::AutoLoadExtension(*client_context, required_extension);
+		}
+
+		FileSystem &fs_after_autoloading = FindFileSystem(path);
+		if (!disabled_file_systems.empty() &&
+		    disabled_file_systems.find(fs_after_autoloading.GetName()) != disabled_file_systems.end()) {
+			throw PermissionException("File system %s has been disabled by configuration",
+			                          fs_after_autoloading.GetName());
+		}
+
+		return fs_after_autoloading;
+	}
+
+	if (!disabled_file_systems.empty() && disabled_file_systems.find(fs.GetName()) != disabled_file_systems.end()) {
+		throw PermissionException("File system %s has been disabled by configuration", fs.GetName());
+	}
+	return fs;
+}
+
 FileSystem &VirtualFileSystem::FindFileSystem(const string &path) {
-	auto &fs = FindFileSystemInternal(path);
+	bool should_autoload = false;
+	auto &fs = FindFileSystemInternal(path, should_autoload);
+	(void)should_autoload; // Here it's irrelevant, we can't autoload in this path
+
 	if (!disabled_file_systems.empty() && disabled_file_systems.find(fs.GetName()) != disabled_file_systems.end()) {
 		throw PermissionException("File system %s has been disabled by configuration", fs.GetName());
 	}
 	return fs;
 }
 
-FileSystem &VirtualFileSystem::FindFileSystemInternal(const string &path) {
+FileSystem &VirtualFileSystem::FindFileSystemInternal(const string &path, bool &should_autoload) {
```

Comment:
How about making this return an `optional_ptr<FileSystem>` and always returning a `nullptr` instead of `default_fs` if none is found